bash-debugger/bdb.sh

```diff
@@ -1,36 +1,75 @@
 #!/bin/bash
+# shellcheck disable=SC2162,SC1090
+
+_e() { echo $'\e['"${1:-}"m ; }
+
+__dbg__usage() {
+    echo "bdb <command> [<command_args>...]"
+    echo
+    echo "Interactive bash script debugger."
+    echo "Runs the <command> with <command_args> and can trace execution steps and/or pause at set breakpoints."
+    echo
+    echo "Commands:"
+    __dbg_commands
+    echo
+}
+__dbg_commands() {
+    echo "  bdb> help           display command list"
+    echo "  bdb> trace          toggle tracing mode [default: off]"
+    echo "  bdb> bl             display breakpoint list"
+    echo "  bdb> ba <expr>      add new breakpoint: pause if <expr> is true, checked every line of code"
+    echo "  bdb> bal <n>        add new breakpoint: pause at line number <n>"
+    echo "  bdb> bae <n>        add new breakpoint: pause at every <n> lines"
+    echo "  bdb> bd <n>         remove the breakpoint number <n> (as in 'bl' command)"
+    echo "  bdb> <command>      run the arbitrary shell command; useful for checking variable values etc."
+    echo
+    echo "Hint: Use ^D instead of Enter to continue the execution."
+}
 
 __dbg__breakpoints=()
 __dbg__trace=2
+__dbg__trap_count=0
 __dbg__trap() {
     local __dbg__cmd __dbg__cmd_args __dbg__set="$(set +o)" \
         __dbg__do_break=false
     set +eu
-    ((__dbg__trace == 1)) \
-        && echo "+(${BASH_SOURCE[1]}:${BASH_LINENO[0]}): $BASH_COMMAND"
-    for __dbg__breakpoint in "${__dbg__breakpoints[@]}"; do
-        eval "$__dbg__breakpoint" && __dbg__do_break=true && break
+    ((__dbg__trap_count++))
+
+    for __dbg__breakpoint_num in $(seq ${#__dbg__breakpoints[@]}); do
```

Comment:
Please declare new variables as `local` to reduce exposure scope and minimize risk of the name conflict.